### DISCUSSION TITLE
valtio for local (component) state

### DISCUSSION THREAD

**Discussion by domagoj-optiopay** [pmndrs/valtio#102] [Q&A]
@dai-shi  Hi, i have a question about using valtio for local component state. i saw that `useLocalProxy` was deprecated recently (like an hour ago 🙃).

Can you maybe explain why was it deprecated and why was it deemed not ideal in the first place, as mentioned [here](https://github.com/pmndrs/valtio/blob/master/src/utils.ts#L15)?

Also are there any plans in the future to officially support using valtio for local state or is that out of scope?

BTW I really like using valtio, even without the support for local state, so in any case - thanks for you work! 👍

**Comment by dai-shi** [comment]
Thanks for opening up a discussion.

> is that out of scope?

It is the out of scope to use it as a primary mean. Is it supported? Yes, I don't mean to break this usage unless there are very unexpected reasons.
For example, can you use redux for local state? Yes, you can. Is it supported? I've never seen the docs describing it.

> why was it deprecated

It was added more or less for experiment. And you can do the same thing at your end.
If you were to use redux as local state, you would use `useRef`.
And, because we have `proxyWithComputed` and any other additional utils patch proxies in the future, we can't support all combinations of them for `useLocalProxyWithFoo`.
So, it's really the matter of how you compose things at your end.

> why was it deemed not ideal in the first place

This is the most important part, and if you understand it, there's nothing wrong with using valtio for local state.
In short, valtio is designed for module state outside react and it requires more work.
So, basically it's less performant than pure useState. Implementation-wise, it depends on useMutableSource (which is for now useSubscription emulation), which does a lot of work to avoid unexpected behavior (tearing) in concurrent mode. And, if it were useState, it could come for free.
In addition, you can't get some CM benefit with module state that would be possible with useState.
So, yeah subscription based mechanism seems over work for local state. But, at the same time it's unavoidable. We need subscriptions to realize valtio api.
In summary, using valtio for local state is valid, but it's not like useState. It's very different.

Actually, if you have some examples that works well with valtio local state, compared to useState, I'm interested in having a look at them.
